chemprop/models/ffn.py

```diff
@@ -0,0 +1,350 @@
+        """
+        Runs the :class:`FFNAtten` on input.
+        :param input: A tuple of atom and bond informations of each molecule.
+        :param constraints: A PyTorch tensor which applies constraint on atomic/bond properties.
+        :param bond_types: A PyTorch tensor storing bond types of each bond determined by RDKit molecules.
+        :return: The output of the :class:`FFNAtten`, a PyTorch tensor containing a list of property predictions.
+        """
+        a_hidden, a_scope, b_hidden, b_scope, b2br = input
+
+        if self.ffn_type == "atom":
+            hidden = a_hidden
+            scope = a_scope
+        elif self.ffn_type == "bond":
+            forward_bond = b_hidden[b2br[:, 0]]
+            backward_bond = b_hidden[b2br[:, 1]]
+            scope = [((start - 1) // 2, size // 2) for start, size in b_scope]
+        else:
+            raise ValueError(f"Unrecognized ffn_type of {self.ffn_type}.")
+
+        if constraints is not None:
+            if self.ffn_type == "atom":
+                output_hidden = self.ffn(hidden)
+                output = self.ffn_readout(output_hidden)
+            elif self.ffn_type == "bond":
+                b_hidden_1 = torch.cat([forward_bond, backward_bond], dim=1)
+                b_hidden_2 = torch.cat([backward_bond, forward_bond], dim=1)
+                output_1 = self.ffn(b_hidden_1)
+                output_2 = self.ffn(b_hidden_2)
+                output_hidden = (output_1 + output_2) / 2
+                output = self.ffn_readout(output_hidden)
+                if bond_types is not None:
+                    output = output + bond_types.reshape(-1, 1)
+
+            u_a = self.weights_readout(output_hidden)
+            constrained_output = []
+            for i, (start, size) in enumerate(scope):
+                if size == 0:
+                    continue
+                else:
+                    u_ai = u_a[start:start+size]
+                    qi = output[start:start+size]
+                    Q = constraints[i]
+
+                    wi = F.softmax(u_ai, dim=0).flatten()
+
+                    qi[:, 0] = qi[:, 0] + wi * (Q - qi[:, 0].sum())
+                    constrained_output.append(qi)
+
+            output = torch.cat(constrained_output, dim=0)
+        else:
+            if self.ffn_type == "atom":
+                output = self.ffn_readout(hidden)
+                output = output[1:]  # remove the first one which is zero padding
+            elif self.ffn_type == "bond":
+                b_hidden_1 = torch.cat([forward_bond, backward_bond], dim=1)
+                b_hidden_2 = torch.cat([backward_bond, forward_bond], dim=1)
+                output_1 = self.ffn_readout(b_hidden_1)
+                output_2 = self.ffn_readout(b_hidden_2)
+                output = (output_1 + output_2) / 2
+                if bond_types is not None:
+                    output = output + bond_types.reshape(-1, 1)
+
+        return output
+
+
+class nn_exp(nn.Module):
```

Comment:
```suggestion
class Exp(nn.Module):
```
classes should be CamelCase